qiskit/providers/ibmq/api/clients/websocket.py

```diff
@@ -161,75 +180,119 @@ def get_job_status(
             WebsocketTimeoutError: if the timeout has been reached.
         """
         url = '{}/jobs/{}/status'.format(self.websocket_url, job_id)
-        websocket = yield from self._connect(url)
 
         original_timeout = timeout
         start_time = time.time()
         attempt_retry = True  # By default, attempt to retry if the websocket connection closes.
+        current_retry_attempt = 0
         last_status = None
+        websocket = None
+
+        while current_retry_attempt <= retries:
+            try:
+                websocket = yield from self._connect(url)
+                # Read messages from the server until the connection is closed or
+                # a timeout has been reached.
+                while True:
+                    try:
+                        with warnings.catch_warnings():
+                            # Suppress websockets deprecation warnings until the fix is available
+                            warnings.filterwarnings("ignore", category=DeprecationWarning)
+                            if timeout:
+                                response_raw = yield from asyncio.wait_for(
+                                    websocket.recv(), timeout=timeout)
+
+                                # Decrease the timeout, with a 5-second grace period.
+                                elapsed_time = time.time() - start_time
+                                timeout = max(5, int(original_timeout - elapsed_time))
+                            else:
+                                response_raw = yield from websocket.recv()
+                        logger.debug('Received message from websocket: %s',
+                                     response_raw)
+
+                        # Successfully received a message, reset retry counter.
+                        current_retry_attempt = 0
+
+                        response = WebsocketMessage.from_bytes(response_raw)
+                        last_status = response.data
+
+                        job_status = response.data.get('status')
+                        if (job_status and
+                                ApiJobStatus(job_status) in API_JOB_FINAL_STATES):
+                            break
```

Comment:
With nested `while` and multiple `break` and `continue`, it'd probably be cleaner to just do a `return` here since the task is done.